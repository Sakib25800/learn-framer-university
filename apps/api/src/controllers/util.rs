use http::request::Parts;
use http::{Extensions, Request};

pub trait RequestPartsExt {
    fn extensions(&self) -> &Extensions;
}

impl RequestPartsExt for Parts {
    fn extensions(&self) -> &Extensions {
        &self.extensions
    }
}

impl<B> RequestPartsExt for Request<B> {
    fn extensions(&self) -> &Extensions {
        self.extensions()
    }
}
