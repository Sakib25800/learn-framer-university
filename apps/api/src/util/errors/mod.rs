//! This module implements several error types and traits.
//!
//! * `util::errors::AppResult` - Failures can be converted into user facing JSON responses. This error type
//!   is more dynamic is box allocated. Low-level errors are typically not converted to user facing errors and most usage
//!   lies within models, controllers, and middleware layers.

use axum::{response::IntoResponse, Extension};
use http::StatusCode;
use std::{
    any::{Any, TypeId},
    borrow::Cow,
    error::Error,
    fmt,
};
use tokio::task::JoinError;
use tracing::*;
use validator::ValidationErrors;

use crate::{email::EmailError, middleware::log_request::ErrorField};
pub use json::{custom, AppErrorResponse};

mod json;

pub type BoxedAppError = Box<dyn AppError>;

pub fn bad_request(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom("Invalid request", StatusCode::BAD_REQUEST, detail)
}

pub fn forbidden(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom("Forbidden", StatusCode::FORBIDDEN, detail)
}

pub fn unauthorized(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom("Unauthorized", StatusCode::UNAUTHORIZED, detail)
}

pub fn not_found(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom("Not found", StatusCode::NOT_FOUND, detail)
}

/// Returns an error with status 503 and the provided description as JSON
pub fn service_unavailable() -> BoxedAppError {
    custom(
        "Service unavailable",
        StatusCode::SERVICE_UNAVAILABLE,
        "Service unavailable",
    )
}

// =============================================================================
// AppError trait

pub trait AppError: Send + fmt::Display + fmt::Debug + 'static {
    /// Generate an HTTP response for the error
    ///
    /// If none is returned, the error will bubble up the middleware stack
    /// where it is eventually logged and a status 500.
    fn response(&self) -> axum::response::Response;

    fn get_type_id(&self) -> TypeId {
        TypeId::of::<Self>()
    }
}

impl dyn AppError {
    pub fn is<T: Any>(&self) -> bool {
        self.get_type_id() == TypeId::of::<T>()
    }
}

impl AppError for BoxedAppError {
    fn response(&self) -> axum::response::Response {
        (**self).response()
    }

    fn get_type_id(&self) -> TypeId {
        (**self).get_type_id()
    }
}

impl IntoResponse for BoxedAppError {
    fn into_response(self) -> axum::response::Response {
        self.response()
    }
}

pub type AppResult<T> = Result<T, BoxedAppError>;

// =============================================================================
// Error impls

impl<E: Error + Send + 'static> AppError for E {
    fn response(&self) -> axum::response::Response {
        error!(error = %self, "Internal Server Error");

        sentry::capture_error(self);

        server_error_response(self.to_string())
    }
}

impl From<sqlx::Error> for BoxedAppError {
    fn from(err: sqlx::Error) -> BoxedAppError {
        match err {
            sqlx::Error::PoolTimedOut => service_unavailable(),
            sqlx::Error::PoolClosed => service_unavailable(),
            sqlx::Error::WorkerCrashed => service_unavailable(),
            sqlx::Error::RowNotFound => not_found("Resource not found"),
            sqlx::Error::Database(db_err) if db_err.is_unique_violation() => {
                bad_request("Resource already exists")
            }
            _ => {
                error!("Database error: {err:?}");
                internal("An internal server error occurred")
            }
        }
    }
}

impl From<prometheus::Error> for BoxedAppError {
    fn from(err: prometheus::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<serde_json::Error> for BoxedAppError {
    fn from(err: serde_json::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<std::io::Error> for BoxedAppError {
    fn from(err: std::io::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<JoinError> for BoxedAppError {
    fn from(err: JoinError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<ValidationErrors> for BoxedAppError {
    fn from(err: ValidationErrors) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<EmailError> for BoxedAppError {
    fn from(error: EmailError) -> Self {
        match error {
            EmailError::Address(error) => Box::new(error),
            EmailError::MessageBuilder(error) => Box::new(error),
            EmailError::Transport(error) => {
                error!(?error, "Failed to send email");
                internal("Failed to send the email")
            }
        }
    }
}

// =============================================================================
// Internal error for use with `chain_error`

#[derive(Debug)]
struct InternalAppError {
    description: String,
}

impl fmt::Display for InternalAppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.description)?;
        Ok(())
    }
}

impl AppError for InternalAppError {
    fn response(&self) -> axum::response::Response {
        error!(error = %self.description, "Internal Server Error");

        sentry::capture_message(&self.description, sentry::Level::Error);

        server_error_response(self.description.to_string())
    }
}

pub fn internal<S: ToString>(error: S) -> BoxedAppError {
    Box::new(InternalAppError {
        description: error.to_string(),
    })
}

fn server_error_response(error: String) -> axum::response::Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Extension(ErrorField(error)),
        "Internal Server Error",
    )
        .into_response()
}
