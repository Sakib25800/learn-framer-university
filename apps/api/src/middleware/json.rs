use crate::util::errors::{bad_request, AppResult, BoxedAppError};
use axum::extract::{FromRequest, Json, Request};
use serde::de::DeserializeOwned;
use validator::{Validate, ValidationErrors, ValidationErrorsKind};

fn generate_user_friendly_message(
    field: &str,
    code: &str,
    params: &std::collections::HashMap<std::borrow::Cow<'static, str>, serde_json::Value>,
) -> String {
    match code {
        "email" => "Invalid email address".to_string(),
        "url" => format!("'{}' is not a valid URL", field),
        "length" => {
            let min = params.get("min").and_then(|v| v.as_u64());
            let max = params.get("max").and_then(|v| v.as_u64());
            let equal = params.get("equal").and_then(|v| v.as_u64());

            match (min, max, equal) {
                (Some(min), Some(max), _) => {
                    format!("'{}' must be between {} and {} characters", field, min, max)
                }
                (Some(min), None, _) => format!("'{}' must be at least {} characters", field, min),
                (None, Some(max), _) => format!("'{}' must not exceed {} characters", field, max),
                (_, _, Some(equal)) => format!("'{}' must be exactly {} characters", field, equal),
                _ => format!("'{}' has an invalid length", field),
            }
        }
        "range" => {
            let min = params.get("min").and_then(|v| v.as_f64());
            let max = params.get("max").and_then(|v| v.as_f64());

            match (min, max) {
                (Some(min), Some(max)) => {
                    format!("'{}' must be between {} and {}", field, min, max)
                }
                (Some(min), None) => format!("'{}' must be at least {}", field, min),
                (None, Some(max)) => format!("'{}' must not exceed {}", field, max),
                _ => format!("'{}' is out of the allowed range", field),
            }
        }
        "must_match" => {
            let other = params
                .get("other")
                .and_then(|v| v.as_str())
                .unwrap_or("other field");
            format!("'{}' must match '{}'", field, other)
        }
        "contains" => format!("'{}' must contain the required value", field),
        "does_not_contain" => format!("'{}' contains a forbidden value", field),
        "regex" => format!("'{}' has an invalid format", field),
        "credit_card" => format!("'{}' is not a valid credit card number", field),
        "required" => format!("'{}' is required", field),
        "non_control_character" => format!("'{}' contains invalid control characters", field),
        _ => format!("'{}' failed validation ({})", field, code),
    }
}

fn get_first_validation_error(errors: ValidationErrors) -> String {
    // Find the first field error.
    for (field, kind) in errors.into_errors() {
        match kind {
            ValidationErrorsKind::Field(errs) => {
                if let Some(err) = errs.into_iter().next() {
                    // Use custom message if available, otherwise generate one.
                    if let Some(msg) = err.message {
                        return msg.to_string();
                    } else {
                        return generate_user_friendly_message(&field, &err.code, &err.params);
                    }
                }
            }
            ValidationErrorsKind::Struct(nested_errors) => {
                // Recursively check nested struct errors.
                return get_first_validation_error(*nested_errors);
            }
            ValidationErrorsKind::List(list_errors) => {
                // Get the first error from the list.
                if let Some((index, nested_errors)) = list_errors.into_iter().next() {
                    let nested_message = get_first_validation_error(*nested_errors);
                    return format!("{} at index {}: {}", field, index, nested_message);
                }
            }
        }
    }

    // Fallback if no specific error is found.
    "Invalid data".to_string()
}

pub struct JsonBody<T>(pub T);

impl<T, S> FromRequest<S> for JsonBody<T>
where
    T: DeserializeOwned + Validate,
    S: Send + Sync,
{
    type Rejection = BoxedAppError;

    async fn from_request(req: Request, state: &S) -> AppResult<Self> {
        let Json(data) = Json::<T>::from_request(req, state)
            .await
            .map_err(|_| bad_request("Invalid JSON"))?;
        data.validate()
            .map_err(|errs| bad_request(get_first_validation_error(errs)))?;
        Ok(JsonBody(data))
    }
}
