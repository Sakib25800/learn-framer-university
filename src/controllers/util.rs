use crate::middleware::app::RequestApp;
use crate::util::errors::{forbidden, AppResult};
use http::request::Parts;
use http::{header, Extensions, HeaderMap, HeaderValue, Request};

/// We don't want to accept authenticated requests that originated from other sites, so this
/// function returns an error if the Origin header doesn't match what we expect "this site" to
/// be: <https://learn.framer.university> in production, or <http://localhost:port/> in development.
pub fn verify_origin(parts: &Parts) -> AppResult<()> {
    let headers = parts.headers();
    let allowed_origins = &parts.app().config.allowed_origins;

    let bad_origin = headers
        .get_all(header::ORIGIN)
        .iter()
        .find(|value| !allowed_origins.contains(value));

    if bad_origin.is_some() {
        return Err(forbidden("Invalid origin header"));
    }

    Ok(())
}

pub trait RequestPartsExt {
    fn headers(&self) -> &HeaderMap<HeaderValue>;
    fn extensions(&self) -> &Extensions;
}

impl RequestPartsExt for Parts {
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        &self.headers
    }
    fn extensions(&self) -> &Extensions {
        &self.extensions
    }
}

impl<B> RequestPartsExt for Request<B> {
    fn headers(&self) -> &HeaderMap<HeaderValue> {
        self.headers()
    }
    fn extensions(&self) -> &Extensions {
        self.extensions()
    }
}
