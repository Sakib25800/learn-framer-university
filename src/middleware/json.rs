use crate::util::errors::{bad_request, AppResult, BoxedAppError};
use axum::extract::{FromRequest, Json, Request};
use serde::de::DeserializeOwned;
use validator::Validate;

pub struct JsonBody<T>(pub T);

impl<T, S> FromRequest<S> for JsonBody<T>
where
    T: DeserializeOwned + Validate,
    S: Send + Sync,
{
    type Rejection = BoxedAppError;

    async fn from_request(req: Request, state: &S) -> AppResult<Self> {
        let Json(data) = Json::<T>::from_request(req, state)
            .await
            .map_err(|err| bad_request(format!("{err}")))?;
        data.validate().map_err(|e| bad_request(format!("{e}")))?;
        Ok(JsonBody(data))
    }
}
