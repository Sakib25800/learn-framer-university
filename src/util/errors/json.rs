use axum::{
    response::{IntoResponse, Response},
    Json,
};
use http::StatusCode;
use serde::Serialize;
use serde_json::json;
use std::{borrow::Cow, fmt};

use super::{AppError, BoxedAppError};

#[derive(Serialize)]
pub struct ErrorResponse {
    title: String,
    status: u16,
    detail: String,
}

/// Generates a response following [RFC 8707 format](https://datatracker.ietf.org/doc/html/rfc7807)
pub fn json_error(title: &str, status: StatusCode, detail: &str) -> Response {
    let json = json!(ErrorResponse {
        title: title.into(),
        status: status.as_u16(),
        detail: detail.into()
    });
    // Wrap the json value with Json
    (status, Json(json)).into_response()
}

// The following structs wrap owned data and provide a custom message to the user

pub fn custom(
    title: impl Into<Cow<'static, str>>,
    status: StatusCode,
    detail: impl Into<Cow<'static, str>>,
) -> BoxedAppError {
    Box::new(CustomApiError {
        status,
        title: title.into(),
        detail: detail.into(),
    })
}

#[derive(Debug, Clone)]
pub struct CustomApiError {
    pub status: StatusCode,
    pub title: Cow<'static, str>,
    pub detail: Cow<'static, str>,
}

impl fmt::Display for CustomApiError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.detail.fmt(f)
    }
}

impl AppError for CustomApiError {
    fn response(&self) -> Response {
        json_error(&self.title, self.status, &self.detail)
    }
}
