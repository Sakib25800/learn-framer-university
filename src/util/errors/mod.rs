//! This module implements several error types and traits.
//!
//! * `util::errors::AppResult` - Failures can be converted into user facing JSON responses. This error type
//!   is more dynamic is box allocated. Low-level errors are typically not converted to user facing errors and most usage
//!   lies within models, controllers, and middleware layers.

use std::{
    any::{Any, TypeId},
    borrow::Cow,
    error::Error,
    fmt,
};

use axum::{response::IntoResponse, Extension};
use diesel::result::{DatabaseErrorKind, Error as DieselError};
use http::StatusCode;
use json::custom;
use tokio::task::JoinError;
use validator::ValidationErrors;

use crate::middleware::log_request::ErrorField;

mod json;

pub type BoxedAppError = Box<dyn AppError>;

pub fn bad_request(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom(StatusCode::BAD_REQUEST, detail)
}

pub fn forbidden(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom(StatusCode::FORBIDDEN, detail)
}

pub fn auth(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom(StatusCode::UNAUTHORIZED, detail)
}

pub fn not_found(detail: impl Into<Cow<'static, str>>) -> BoxedAppError {
    custom(StatusCode::NOT_FOUND, detail)
}

/// Returns an error with status 503 and the provided description as JSON
pub fn service_unavailable() -> BoxedAppError {
    custom(StatusCode::SERVICE_UNAVAILABLE, "Service unavailable")
}

// =============================================================================
// AppError trait

pub trait AppError: Send + fmt::Display + fmt::Debug + 'static {
    /// Generate an HTTP response for the error
    ///
    /// If none is returned, the error will bublbe up the middleware stack
    /// where it is eventually logged and a status 500.
    fn response(&self) -> axum::response::Response;

    fn get_type_id(&self) -> TypeId {
        TypeId::of::<Self>()
    }
}

impl dyn AppError {
    pub fn is<T: Any>(&self) -> bool {
        self.get_type_id() == TypeId::of::<T>()
    }
}

impl AppError for BoxedAppError {
    fn response(&self) -> axum::response::Response {
        (**self).response()
    }

    fn get_type_id(&self) -> TypeId {
        (**self).get_type_id()
    }
}

impl IntoResponse for BoxedAppError {
    fn into_response(self) -> axum::response::Response {
        self.response()
    }
}

pub type AppResult<T> = Result<T, BoxedAppError>;

// =============================================================================
// Error impls

impl<E: Error + Send + 'static> AppError for E {
    fn response(&self) -> axum::response::Response {
        error!(error = %self, "Internal Server Error");

        sentry::capture_error(self);

        server_error_response(self.to_string())
    }
}

impl From<diesel::ConnectionError> for BoxedAppError {
    fn from(err: diesel::ConnectionError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<DieselError> for BoxedAppError {
    fn from(err: DieselError) -> BoxedAppError {
        match err {
            DieselError::NotFound => not_found(""),
            DieselError::DatabaseError(DatabaseErrorKind::ClosedConnection, _) => {
                service_unavailable()
            }
            _ => Box::new(err),
        }
    }
}

impl From<diesel_async::pooled_connection::deadpool::PoolError> for BoxedAppError {
    fn from(err: diesel_async::pooled_connection::deadpool::PoolError) -> BoxedAppError {
        error!("Database pool error: {err}");
        service_unavailable()
    }
}

impl From<prometheus::Error> for BoxedAppError {
    fn from(err: prometheus::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<serde_json::Error> for BoxedAppError {
    fn from(err: serde_json::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<std::io::Error> for BoxedAppError {
    fn from(err: std::io::Error) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<JoinError> for BoxedAppError {
    fn from(err: JoinError) -> BoxedAppError {
        Box::new(err)
    }
}

impl From<ValidationErrors> for BoxedAppError {
    fn from(err: ValidationErrors) -> BoxedAppError {
        Box::new(err)
    }
}

// =============================================================================
// Internal error for use with `chain_error`

#[derive(Debug)]
struct InternalAppError {
    description: String,
}

impl fmt::Display for InternalAppError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.description)?;
        Ok(())
    }
}

impl AppError for InternalAppError {
    fn response(&self) -> axum::response::Response {
        error!(error = %self.description, "Internal Server Error");

        sentry::capture_message(&self.description, sentry::Level::Error);

        server_error_response(self.description.to_string())
    }
}

pub fn internal<S: ToString>(error: S) -> BoxedAppError {
    Box::new(InternalAppError {
        description: error.to_string(),
    })
}

fn server_error_response(error: String) -> axum::response::Response {
    (
        StatusCode::INTERNAL_SERVER_ERROR,
        Extension(ErrorField(error)),
        "Internal Server Error",
    )
        .into_response()
}

#[cfg(test)]
mod tests {
    use super::*;
    use diesel::result::Error as DieselError;
    use http::StatusCode;

    #[test]
    fn http_error_responses() {
        use crate::serde::de::Error;

        // Types for handling common error status codes
        assert_eq!(bad_request("").response().status(), StatusCode::BAD_REQUEST);
        assert_eq!(forbidden("").response().status(), StatusCode::FORBIDDEN);
        assert_eq!(
            BoxedAppError::from(DieselError::NotFound)
                .response()
                .status(),
            StatusCode::NOT_FOUND
        );
        assert_eq!(not_found("").response().status(), StatusCode::NOT_FOUND);

        // All other error types are converted to internal server errors
        assert_eq!(
            internal("").response().status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(serde_json::Error::custom("ExpectedColon"))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
        assert_eq!(
            BoxedAppError::from(std::io::Error::new(::std::io::ErrorKind::Other, ""))
                .response()
                .status(),
            StatusCode::INTERNAL_SERVER_ERROR
        );
    }
}
