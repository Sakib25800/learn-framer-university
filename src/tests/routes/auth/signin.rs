use crate::tests::util::{test_app::TestApp, RequestHelper};
use http::StatusCode;
use insta::assert_snapshot;
use serde_json::json;

#[tokio::test(flavor = "multi_thread")]
async fn signin_should_succeed_with_valid_email() {
    let (app, anon) = TestApp::init().empty().await;

    let sign_in_response = anon
        .post("/v1/auth/signin")
        .json(&json!({
            "email": "new_user@example.com"
        }))
        .await;

    sign_in_response.assert_status(StatusCode::OK);
    sign_in_response.assert_json(&json!({
        "message": "We've sent you an email",
    }));

    let emails = app.emails().await;

    assert_snapshot!(app.emails_snapshot().await);

    // Retrieve the continue token
    let continue_token = extract_token_from_signin_email(&emails);

    // Continue with token
    let continue_path = format!("/v1/auth/continue/{continue_token}");
    let continue_response = anon.get(&continue_path).await;

    continue_response.assert_status_ok();

    let json_continue_response = continue_response.json::<serde_json::Value>();

    match json_continue_response {
        serde_json::Value::Object(map) => {
            assert!(map.contains_key("access_token"));
            assert!(map.contains_key("refresh_token"));
        }
        _ => panic!("Expected JSON object"),
    }
}

#[tokio::test(flavor = "multi_thread")]
async fn signin_should_reject_with_invalid_token() {
    let (_, anon) = TestApp::init().empty().await;

    let continue_path = "/v1/auth/continue/invalid_token";
    let response = anon.get(continue_path).await;

    response.assert_status_unauthorized();
    response.assert_json(&json!({
        "title": "Unauthorized",
        "detail": "Invalid verification token",
        "status": 401
    }));
}

#[tokio::test(flavor = "multi_thread")]
async fn signin_should_reject_invalid_email_format() {
    let (app, anon) = TestApp::init().empty().await;

    let response = anon
        .post("/v1/auth/signin")
        .json(&json!({
            "email": "not-an-email"
        }))
        .await;

    response.assert_status_bad_request();

    assert!(app.emails().await.is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn signin_should_reject_when_email_missing() {
    let (app, anon) = TestApp::init().empty().await;

    let response = anon.post("/v1/auth/signin").json(&json!({})).await;

    response.assert_status_bad_request();

    assert!(app.emails().await.is_empty());
}

#[tokio::test(flavor = "multi_thread")]
async fn signin_should_work_for_existing_user() {
    let (app, anon, user) = TestApp::init().with_user().await;

    let response = anon
        .post("/v1/auth/signin")
        .json(&json!({
            "email": user.as_model().email
        }))
        .await;

    response.assert_status_ok();
    response.assert_json(&json!({
        "message": "We've sent you an email"
    }));

    let emails = app.emails().await;

    assert_eq!(emails.len(), 1);
}

fn extract_token_from_signin_email(emails: &[String]) -> String {
    let body = emails
        .iter()
        .find(|m| m.contains("Subject: Activation link for Framer University"))
        .expect("Missing email");

    let after_prefix = body
        .split("/continue/")
        .nth(1)
        .expect("Couldn't find token start");

    let token = after_prefix
        .split_whitespace()
        .next()
        .expect("Couldn't find token end");

    token.to_string()
}
