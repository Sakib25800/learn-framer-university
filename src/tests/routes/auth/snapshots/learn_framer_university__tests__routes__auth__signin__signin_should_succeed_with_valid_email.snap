---
source: src/tests/routes/auth/signin.rs
expression: app.emails_snapshot().await
---
To: new_user@example.com
From: learn.framer.university <noreply@learn.framer.university>
Subject: Framer University: Please confirm your email address
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

Hey there! Welcome to Framer University.
Please click the link below to sign in: http://learn.framer.university/auth=
/continue/[continue-token]
